miniwin/src/d3drm/d3drmviewport.cpp

```diff
@@ -99,39 +96,58 @@ static void ComputeFrameWorldMatrix(IDirect3DRMFrame* frame, D3DRMMATRIX4D out)
 	memcpy(out, acc, sizeof(acc));
 }
 
+inline D3DVECTOR CrossProduct(const D3DVECTOR& a, const D3DVECTOR& b)
+{
+	return {a.y * b.z - a.z * b.y, a.z * b.x - a.x * b.z, a.x * b.y - a.y * b.x};
+}
+
+inline D3DVECTOR Normalize(const D3DVECTOR& v)
+{
+	float len = sqrtf(v.x * v.x + v.y * v.y + v.z * v.z);
+	if (len > 0.0f) {
+		float invLen = 1.0f / len;
+		return {v.x * invLen, v.y * invLen, v.z * invLen};
+	}
+	return {0, 0, 0};
+}
+
 D3DVECTOR ComputeTriangleNormal(const D3DVECTOR& v0, const D3DVECTOR& v1, const D3DVECTOR& v2)
 {
 	D3DVECTOR u = {v1.x - v0.x, v1.y - v0.y, v1.z - v0.z};
 	D3DVECTOR v = {v2.x - v0.x, v2.y - v0.y, v2.z - v0.z};
-	D3DVECTOR normal = {u.y * v.z - u.z * v.y, u.z * v.x - u.x * v.z, u.x * v.y - u.y * v.x};
-	float len = std::sqrt(normal.x * normal.x + normal.y * normal.y + normal.z * normal.z);
-	if (len > 0.0f) {
-		normal.x /= len;
-		normal.y /= len;
-		normal.z /= len;
-	}
+	D3DVECTOR normal = CrossProduct(u, v);
+	normal = Normalize(normal);
 
 	return normal;
 }
 
-inline D3DVECTOR TransformPoint(const D3DVECTOR& p, D3DRMMATRIX4D& m)
+inline D3DVECTOR TransformNormal(const D3DVECTOR& v, const Matrix3x3& m)
+{
+	return {
+		v.x * m[0][0] + v.y * m[1][0] + v.z * m[2][0],
+		v.x * m[0][1] + v.y * m[1][1] + v.z * m[2][1],
+		v.x * m[0][2] + v.y * m[1][2] + v.z * m[2][2]
+	};
+}
+
+inline D3DVECTOR TransformPoint(const D3DVECTOR& v, const D3DRMMATRIX4D& m)
```

Comment:
little bikeshed, but `p` for point makes more sense here.
A point is a position with `w` component != 0.
A vector is a directtion with `w` component= = 0.